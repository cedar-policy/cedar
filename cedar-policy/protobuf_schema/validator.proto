//
// Copyright Cedar Contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

syntax = "proto3";
package cedar_policy_validator;

import "core.proto";

// the protobuf Schema message describes a complete schema.
message Schema {
    // TODO: this need not be a map at all, since `EntityDecl` contains the `name` as well.
    // It can be just `repeated EntityDecl`.
    repeated EntityTypeToEntityDeclMap entity_decls = 1;
    // TODO: this need not be a map at all, since `ActionDecl` contains the `name` as well.
    // It can be just `repeated ActionDecl`.
    repeated EntityUidToActionDeclMap action_decls = 2;
}

// This `message` with `key` and `value`, rather than a `map`, since messages can't be dictionary keys
message EntityTypeToEntityDeclMap {
    cedar_policy_core.Name key = 1;
    EntityDecl value = 2;
}

// This `message` with `key` and `value`, rather than a `map`, since messages can't be dictionary keys
message EntityUidToActionDeclMap {
    cedar_policy_core.EntityUid key = 1;
    ActionDecl value = 2;
}

// the protobuf EntityDecl message contains all of the schema's
// information about a single entity type.
message EntityDecl {
    cedar_policy_core.Name name = 1;
    repeated cedar_policy_core.Name descendants = 2;
    map<string, AttributeType> attributes = 3;
    OpenTag open_attributes = 4;
    optional Type tags = 5;
    repeated string enum_choices = 6;
}

// the protobuf ActionDecl message contains all of the schema's
// information about a single action.
message ActionDecl {
    cedar_policy_core.EntityUid name = 1;
    repeated cedar_policy_core.EntityUid descendants = 3;
    Type context = 4;
    repeated cedar_policy_core.Name principal_types = 7;
    repeated cedar_policy_core.Name resource_types = 8;
}

message Type {
    oneof data {
        Ty ty = 1;
        Type set_type = 2;
        EntityRecordKind entityOrRecord = 3;
        cedar_policy_core.Name name = 4;
    }

    enum Ty {
        Never = 0;
        True = 1;
        False = 2;
        EmptySetType = 3;
        Bool = 4;
        String = 5;
        Long = 6;
    }
}

message EntityRecordKind {
    oneof data {
        AnyEntity any_entity = 1;
        Record record = 2;
        cedar_policy_core.Name entity = 3;
        ActionEntity actionEntity = 4;
    }

    // Zero-arity constructors represented as enums with only one member
    enum AnyEntity {
        // the one option for the enum
        unit = 0;
    }
    message Record {
        map<string, AttributeType> attrs = 1;
        OpenTag open_attributes = 2;
    }
    message ActionEntity {
        cedar_policy_core.Name name = 1;
        map<string, AttributeType> attrs = 2;
    }
}

enum OpenTag {
    OpenAttributes = 0;
    ClosedAttributes = 1;
}

message AttributeType {
    Type attr_type = 1;
    bool is_required = 2;
}

enum ValidationMode {
    Strict = 0;
    Permissive = 1;
}
