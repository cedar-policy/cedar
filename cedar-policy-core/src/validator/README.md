# Cedar Policy Validator

This submodule contains the validator for Cedar policies.

This submodule exposes low-level and advanced APIs for Cedar policy validation.
Anyone simply wanting to use Cedar from a Rust client (e.g., to validate that
policies do not contain run time type errors) should use
[`cedar-policy`](../../../cedar-policy) instead.

## Documentation

Generated documentation for the latest version can be accessed
[on docs.rs](https://docs.rs/cedar-policy-core/latest/cedar_policy_core/validator/index.html).
