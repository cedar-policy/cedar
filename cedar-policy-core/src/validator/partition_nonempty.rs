use itertools::Itertools;
use nonempty::NonEmpty;

/// A trait for partitioning a collection of `Result`s into a collection of `Ok` values or a `NonEmpty` of `Err` values.
pub(crate) trait PartitionNonEmpty<T, E> {
    fn partition_nonempty<C>(self) -> std::result::Result<C, NonEmpty<E>>
    where
        C: Default + Extend<T>;
}

impl<I, T, E> PartitionNonEmpty<T, E> for I
where
    I: Iterator<Item = Result<T, E>>,
{
    fn partition_nonempty<C>(self) -> Result<C, NonEmpty<E>>
    where
        C: Default + Extend<T>,
    {
        let (oks, errs): (_, Vec<_>) = self.partition_result();

        if let Some(errs) = NonEmpty::from_vec(errs) {
            Err(errs)
        } else {
            Ok(oks)
        }
    }
}
