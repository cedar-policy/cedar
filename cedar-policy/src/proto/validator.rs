/*
 * Copyright Cedar Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#![allow(clippy::use_self)]

use super::models;
use cedar_policy_core::ast;
use cedar_policy_validator::types;
use nonempty::NonEmpty;
use smol_str::SmolStr;
use std::collections::{BTreeMap, HashMap};

impl From<&cedar_policy_validator::ValidatorSchema> for models::Schema {
    fn from(v: &cedar_policy_validator::ValidatorSchema) -> Self {
        Self {
            entity_decls: v.entity_types().map(models::EntityDecl::from).collect(),
            action_decls: v.action_ids().map(models::ActionDecl::from).collect(),
        }
    }
}

impl From<&models::Schema> for cedar_policy_validator::ValidatorSchema {
    // PANIC SAFETY: experimental feature
    #[allow(clippy::expect_used)]
    fn from(v: &models::Schema) -> Self {
        Self::new(
            v.entity_decls
                .iter()
                .map(cedar_policy_validator::ValidatorEntityType::from),
            v.action_decls
                .iter()
                .map(cedar_policy_validator::ValidatorActionId::from),
        )
    }
}

impl From<&cedar_policy_validator::ValidationMode> for models::ValidationMode {
    // PANIC SAFETY: experimental feature
    #[allow(clippy::unimplemented)]
    fn from(v: &cedar_policy_validator::ValidationMode) -> Self {
        match v {
            cedar_policy_validator::ValidationMode::Strict => models::ValidationMode::Strict,
            cedar_policy_validator::ValidationMode::Permissive => {
                models::ValidationMode::Permissive
            }
            #[cfg(feature = "partial-validate")]
            cedar_policy_validator::ValidationMode::Partial => unimplemented!(),
        }
    }
}

impl From<&models::ValidationMode> for cedar_policy_validator::ValidationMode {
    fn from(v: &models::ValidationMode) -> Self {
        match v {
            models::ValidationMode::Strict => cedar_policy_validator::ValidationMode::Strict,
            models::ValidationMode::Permissive => {
                cedar_policy_validator::ValidationMode::Permissive
            }
        }
    }
}

// PANIC SAFETY: experimental feature
#[allow(clippy::fallible_impl_from)]
impl From<&cedar_policy_validator::ValidatorActionId> for models::ActionDecl {
    // PANIC SAFETY: experimental feature
    #[allow(clippy::panic)]
    fn from(v: &cedar_policy_validator::ValidatorActionId) -> Self {
        debug_assert_eq!(
            v.attribute_types().keys().collect::<Vec<&SmolStr>>(),
            Vec::<&SmolStr>::new(),
            "action attributes are not currently supported in protobuf"
        );
        debug_assert_eq!(
            v.attributes().collect::<Vec<_>>(),
            vec![],
            "action attributes are not currently supported in protobuf"
        );
        let ctx_attrs = match v.context() {
            types::Type::EntityOrRecord(types::EntityRecordKind::Record {
                attrs,
                open_attributes: types::OpenTag::ClosedAttributes,
            }) => attrs,
            ty => panic!("expected context to be a closed record, but got {ty:?}"),
        };
        Self {
            name: Some(models::EntityUid::from(v.name())),
            principal_types: v.applies_to_principals().map(models::Name::from).collect(),
            resource_types: v.applies_to_resources().map(models::Name::from).collect(),
            descendants: v.descendants().map(models::EntityUid::from).collect(),
            context: attributes_to_model(ctx_attrs),
        }
    }
}

impl From<&models::ActionDecl> for cedar_policy_validator::ValidatorActionId {
    // PANIC SAFETY: experimental feature
    #[allow(clippy::expect_used)]
    fn from(v: &models::ActionDecl) -> Self {
        Self::new(
            ast::EntityUID::from(v.name.as_ref().expect("name field should exist")),
            v.principal_types.iter().map(ast::EntityType::from),
            v.resource_types.iter().map(ast::EntityType::from),
            v.descendants.iter().map(ast::EntityUID::from),
            types::Type::EntityOrRecord(types::EntityRecordKind::Record {
                attrs: model_to_attributes(&v.context),
                open_attributes: types::OpenTag::default(),
            }),
            // protobuf formats do not include action attributes, so we
            // translate into a `ValidatorActionId` with no action attributes
            types::Attributes::with_attributes([]),
            BTreeMap::new(),
            None,
        )
    }
}

impl From<&cedar_policy_validator::ValidatorEntityType> for models::EntityDecl {
    fn from(v: &cedar_policy_validator::ValidatorEntityType) -> Self {
        let name = Some(models::Name::from(v.name()));
        let descendants = v.descendants.iter().map(models::Name::from).collect();
        let attributes = attributes_to_model(v.attributes());
        let tags = v.tag_type().map(models::Type::from);
        match &v.kind {
            cedar_policy_validator::ValidatorEntityTypeKind::Standard(_) => Self {
                name,
                descendants,
                attributes,
                tags,
                enum_choices: vec![],
            },
            cedar_policy_validator::ValidatorEntityTypeKind::Enum(enum_choices) => Self {
                name,
                descendants,
                attributes,
                tags,
                enum_choices: enum_choices.into_iter().map(ToString::to_string).collect(),
            },
        }
    }
}

impl From<&models::EntityDecl> for cedar_policy_validator::ValidatorEntityType {
    // PANIC SAFETY: experimental feature
    #[allow(clippy::expect_used)]
    fn from(v: &models::EntityDecl) -> Self {
        let name = ast::EntityType::from(v.name.as_ref().expect("name field should exist"));
        let descendants = v.descendants.iter().map(ast::EntityType::from);
        match NonEmpty::collect(v.enum_choices.iter().map(SmolStr::from)) {
            // `enum_choices` is empty, so `v` represents a standard entity type
            None => Self::new_standard(
                name,
                descendants,
                model_to_attributes(&v.attributes),
                types::OpenTag::default(),
                v.tags.as_ref().map(types::Type::from),
                None,
            ),
            Some(enum_choices) => {
                // `enum_choices` is not empty, so `v` represents an enumerated entity type.
                // enumerated entity types must have no attributes or tags.
                assert_eq!(&v.attributes, &HashMap::new());
                assert_eq!(&v.tags, &None);
                Self::new_enum(name.clone(), descendants, enum_choices, name.loc().cloned())
            }
        }
    }
}

impl From<&models::Type> for types::Type {
    // PANIC SAFETY: experimental feature
    #[allow(clippy::expect_used)]
    fn from(v: &models::Type) -> Self {
        match v.data.as_ref().expect("data field should exist") {
            models::r#type::Data::Prim(vt) => {
                match models::r#type::Prim::try_from(vt.to_owned()).expect("decode should succeed")
                {
                    models::r#type::Prim::Bool => types::Type::primitive_boolean(),
                    models::r#type::Prim::String => types::Type::primitive_string(),
                    models::r#type::Prim::Long => types::Type::primitive_long(),
                }
            }
            models::r#type::Data::SetElem(elty) => types::Type::Set {
                element_type: Some(Box::new(types::Type::from(elty.as_ref()))),
            },
            models::r#type::Data::Entity(e) => {
                types::Type::EntityOrRecord(types::EntityRecordKind::Entity(
                    types::EntityLUB::single_entity(ast::EntityType::from(e)),
                ))
            }
            models::r#type::Data::Record(r) => {
                types::Type::EntityOrRecord(types::EntityRecordKind::Record {
                    attrs: model_to_attributes(&r.attrs),
                    open_attributes: types::OpenTag::default(),
                })
            }
            models::r#type::Data::Ext(name) => types::Type::ExtensionType {
                name: ast::Name::from(name),
            },
        }
    }
}

// PANIC SAFETY: experimental feature
#[allow(clippy::fallible_impl_from)]
impl From<&types::Type> for models::Type {
    // PANIC SAFETY: experimental feature
    #[allow(clippy::expect_used, clippy::panic)]
    fn from(v: &types::Type) -> Self {
        match v {
            types::Type::Never => panic!("can't encode Never type in protobuf; Never should never appear in a Schema"),
            types::Type::True | types::Type::False => panic!("can't encode singleton boolean type in protobuf; singleton boolean types should never appear in a Schema"),
            types::Type::Primitive { primitive_type } => match primitive_type {
                types::Primitive::Bool => Self {
                    data: Some(models::r#type::Data::Prim(models::r#type::Prim::Bool.into())),
                },
                types::Primitive::Long => Self {
                    data: Some(models::r#type::Data::Prim(models::r#type::Prim::Long.into())),
                },
                types::Primitive::String => Self {
                    data: Some(models::r#type::Data::Prim(models::r#type::Prim::String.into())),
                },
            },
            types::Type::Set { element_type } => Self {
                data: Some(models::r#type::Data::SetElem(Box::new(models::Type::from(
                    element_type
                        .as_ref()
                        .expect("can't encode Set without element type in protobuf; Set-without-element-type should never appear in a Schema")
                        .as_ref(),
                )))),
            },
            types::Type::EntityOrRecord(types::EntityRecordKind::Entity(lub)) => Self {
                data: Some(models::r#type::Data::Entity(models::Name::from(lub.get_single_entity().expect("can't encode non-singleton LUB in protobuf; non-singleton LUB types should never appear in a Schema").as_ref()))),
            },
            types::Type::EntityOrRecord(types::EntityRecordKind::Record { attrs, open_attributes }) => {
                assert_eq!(open_attributes, &types::OpenTag::ClosedAttributes, "can't encode open record in protobuf");
                Self {
                    data: Some(models::r#type::Data::Record(models::r#type::Record { attrs: attributes_to_model(attrs) })),
                }
            }
            types::Type::EntityOrRecord(types::EntityRecordKind::ActionEntity { name, attrs }) => {
                debug_assert_eq!(attrs.keys().collect::<Vec<&SmolStr>>(), Vec::<&SmolStr>::new(), "can't encode action attributes in protobuf");
                Self {
                    data: Some(models::r#type::Data::Entity(models::Name::from(name.as_ref()))),
                }
            }
            types::Type::EntityOrRecord(types::EntityRecordKind::AnyEntity) => panic!("can't encode AnyEntity type in protobuf; AnyEntity should never appear in a Schema"),
            types::Type::ExtensionType { name } => Self {
                data: Some(models::r#type::Data::Ext(models::Name::from(name))),
            },
        }
    }
}

fn model_to_attributes(v: &HashMap<String, models::AttributeType>) -> types::Attributes {
    types::Attributes::with_attributes(v.iter().map(|(k, v)| (k.into(), v.into())))
}

fn attributes_to_model(v: &types::Attributes) -> HashMap<String, models::AttributeType> {
    v.iter()
        .map(|(k, v)| (k.to_string(), models::AttributeType::from(v)))
        .collect()
}

impl From<&models::AttributeType> for types::AttributeType {
    // PANIC SAFETY: experimental feature
    #[allow(clippy::expect_used)]
    fn from(v: &models::AttributeType) -> Self {
        Self {
            attr_type: types::Type::from(
                v.attr_type.as_ref().expect("attr_type field should exist"),
            ),
            is_required: v.is_required,
        }
    }
}

impl From<&types::AttributeType> for models::AttributeType {
    fn from(v: &types::AttributeType) -> Self {
        Self {
            attr_type: Some(models::Type::from(&v.attr_type)),
            is_required: v.is_required,
        }
    }
}
