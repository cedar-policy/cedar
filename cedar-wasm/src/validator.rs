use cedar_policy::frontend::{utils::InterfaceResult, validate::json_validate};
use wasm_bindgen::prelude::*;

#[wasm_bindgen(js_name = "validate")]
pub fn wasm_validate(input: &str) -> InterfaceResult {
    json_validate(input)
}
