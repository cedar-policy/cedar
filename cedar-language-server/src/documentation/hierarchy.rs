/* * Copyright Cedar Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use std::borrow::Cow;

use cedar_policy_core::validator::ValidatorSchema;

use crate::{impl_documentation_from_markdown_file, policy::cedar::CedarTypeKind};

use super::ToDocumentationString;

#[derive(Default)]
pub(crate) struct SetDocumentation(Option<CedarTypeKind>);

impl SetDocumentation {
    pub(crate) fn new(set_type: CedarTypeKind) -> Self {
        Self(Some(set_type))
    }
}

impl ToDocumentationString for SetDocumentation {
    fn to_documentation_string(&self, _schema: Option<&ValidatorSchema>) -> Cow<'static, str> {
        let content = include_str!("markdown/hierarchy/set.md");
        match &self.0 {
            Some(elem_type) => content.replace('?', &elem_type.to_string()).into(),
            None => content.into(),
        }
    }
}

impl_documentation_from_markdown_file!(InDocumentation, "markdown/hierarchy/in.md");
impl_documentation_from_markdown_file!(HasDocumentation, "markdown/hierarchy/has.md");
impl_documentation_from_markdown_file!(IsDocumentation, "markdown/hierarchy/is.md");
impl_documentation_from_markdown_file!(ContainsDocumentation, "markdown/hierarchy/contains.md");
impl_documentation_from_markdown_file!(
    ContainsAllDocumentation,
    "markdown/hierarchy/contains_all.md"
);
impl_documentation_from_markdown_file!(
    ContainsAnyDocumentation,
    "markdown/hierarchy/contains_any.md"
);
impl_documentation_from_markdown_file!(IsEmptyDocumentation, "markdown/hierarchy/is_empty.md");

#[cfg(test)]
mod test {
    use std::{collections::BTreeMap, sync::Arc};

    use insta::assert_snapshot;
    use smol_str::ToSmolStr;

    use crate::{
        documentation::{SetDocumentation, ToDocumentationString},
        policy::cedar::{Attribute, CedarTypeKind, Record},
    };

    #[test]
    fn test_set() {
        let docs = SetDocumentation::new(CedarTypeKind::String);
        assert_snapshot!(docs.to_documentation_string(None));
        let docs = SetDocumentation::new(CedarTypeKind::Record(Record {
            attrs: Arc::new(BTreeMap::from([(
                "foo".to_smolstr(),
                Attribute::new("foo".to_smolstr(), true, Some(CedarTypeKind::Bool)),
            )])),
        }));
        assert_snapshot!(docs.to_documentation_string(None));
    }
}
