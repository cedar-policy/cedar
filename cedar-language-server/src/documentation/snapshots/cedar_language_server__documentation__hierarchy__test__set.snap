---
source: cedar-language-server/src/documentation/hierarchy.rs
expression: docs.to_documentation_string(None)
---
# Set Type

```cedarschema
Set<string>
```

A collection type that contains elements of type `string`.
